docs/signature-layout.md

```diff
@@ -0,0 +1,35 @@
+# Signature File Layout
+
+A common file layout for storing and serving signatures provides a consistent way to reference image signatures. Signatures on a filesystem or a web server shall use this common layout. Signatures stored in a REST API are not required to use this common layout.
+
+## Specification
+
+This specification relies on [RFC3986](https://tools.ietf.org/html/rfc3986), focusing on defining a [path component](https://tools.ietf.org/html/rfc3986#section-3.3) to compose a concise URI reference to a signature.
+
+**SCHEME[AUTHORITY]/PATH_PREFIX/IMAGE@MANIFEST_DIGEST/signature-INT**
+
+**Definitions**
+
+* **SCHEME**: URI scheme per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1), e.g. **file://** or **https://**
+* **AUTHORITY**: An optional authority reference per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2), e.g. **example.com**
+* **PATH_PREFIX**: An arbitrary base path to the image component
+* **IMAGE**: The name of the image per [v2 API](https://docs.docker.com/registry/spec/api/#/overview). This would typically take the form of registry/repository/image but is not required to have exactly three parts. There is no requirement to include a **:PORT** component but it should be included if part of the image reference.
```

Comment:
> @mtrmac I have a feeling that you have a particular scenario in mind that involves many users, many keys, and lots of access restrictions. Perhaps it would help to define those use cases directly and explicitly, and then we could more easily see which designs would accommodate them.

Fair enough, I suppose I should address this directly. Think a multi-tenant PaaS like OpenShift, which hosts various customers, quite possibly competitors; or a multi-tenant docker registry (“container app store”). Then the _default_ expectation is that each tenant is completely private, even though they actually use a shared infrastructure; OTOH for ease of use/setup, and scaling/load-balancing, or as an _essential purpose_ of the “app store”, it is advantageous to provide a single registry service (= a single cluster of servers, a single host name) for all of them, and to separate the tenants only via authentication and access controls.

Then it’s not a matter of “well, if we disclose signatures of one tenant to another, is that so bad?” as much as “if it is easy to separate them, why would we ever mix them together in the first place?”. Presumably each tenant’s images are protected by some kind of access control, both for read and write; then it would be natural to expect that the signatures are protected exactly the same, using exactly the same ACLs applied to exactly the same repo names and repo namespaces.

---

And there is a specific bad failure case: removing signatures.

Suppose tenants A and B both upload a copy of `rhel/rhel7`, with the original RH signatures, as their base image, and then each tenant signs it using their own keyA and keyB to mark it as an approved base image. Fine. With a shared namespace tenant A would see keyB and vice versa, and would learn about the existence of B and about the name of the image in B’s namespace, which is already pretty undesirable, but let’s accept that.

Then, A sets up a policy “only images signed by keyA are acceptable”, while B sets up a policy “only images signed _both_ by RH and keyB are acceptable”. OK, things still work.

Then A decides to remove the RH signature from its copy of `rhel/rhel7`. Well, why couldn’t they, it’s their copy of an image, living in their own namespace, and A doesn’t need RH’s signature any more, right? But when the signatures are shared between A and B, B’s copy of the image also sees the RH signature removed, and B’s copy of the image becomes unacceptable per B’s policy.

And this can of course happen not only for RH’s signatures, but for third-party signatures, say by an auditor’s keyC. I can’t see a very natural way for the multi-tenant registry to decide which signature removals should be allowed and which should be accepted; it just becomes a horrible mess. In a first version, the system would probably prohibit removing signatures at all, and every customer would see 50 (or 10000?) signatures on the shared `rhel/rhel7` image which they don’t care about and which would never be removed.